src/pudl/output/ferc1.py

```diff
@@ -1399,3 +1402,341 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
+                f"Exploded metadataframes must be indexed by {idx_cols}, but found "
+                f"{df.index.names=}"
+            )
+        if not df.index.is_unique:
+            raise AssertionError("Found non-unique index in exploded metadata.")
+
+    @classmethod
+    def from_exploded_meta(
+        cls,
+        source_table: str,
+        xbrl_factoid: str,
+        exploded_meta: pd.DataFrame,
+        weight: float = np.nan,
+        tags_df: pd.DataFrame | None = None,
+        propagate_weights: bool = True,
+    ) -> "XbrlCalculationTreeFerc1":
+        """Construct a complete calculation tree based on exploded metadata.
+
+        Args:
+            source_table: The table within which we are searching for the XBRL fact,
+                required as many appear in multiple tables.
+            xbrl_factoid: Equivalent to the "name" field in the calculation metadata.
+                Potentially renamed from the originally reported XBRL value.
+            exploded_meta: A dataframe containing metadata for all of the facts
+                referenced by the exploded tables. Must be indexed by the columns
+                (table_name, xbrl_factoid) and the index must be unique.
+            weight: The weight associated with this node in its calculation. Must be
+                passed in because it is only available when we have access to the
+                calculation metadata. Defaults to 1.0 so that we can create the root
+                node in a tree (which is not embedded in any calculation) without having
+                to specify the weight.
+
+        Returns:
+            The root node of a calculation tree, potentially referencing child nodes.
+        """
+        cls._check_index(exploded_meta)
+
+        # Index to look up the particular factoid we're building the node for:
+        idx = (source_table, xbrl_factoid)
+        # Read in the calculations (if any) that define the node:
+        calculations = json.loads(exploded_meta.at[idx, "calculations"])
+        children = []
+        for calc in calculations:
+            if len(calc["source_tables"]) != 1:
+                raise AssertionError(
+                    "Generating the calculation tree for exploded tables requires all "
+                    "xbrl_factoids to have a unique source table, but found "
+                    f"{calc=}"
+                )
+            children.append(
+                XbrlCalculationTreeFerc1.from_exploded_meta(
+                    weight=calc["weight"],
+                    xbrl_factoid=calc["name"],
+                    source_table=calc["source_tables"][0],
+                    exploded_meta=exploded_meta,
+                )
+            )
+
+        tree = XbrlCalculationTreeFerc1(
+            source_table=source_table,
+            xbrl_factoid=xbrl_factoid,
+            weight=weight,
+            xbrl_factoid_original=exploded_meta.at[idx, "xbrl_factoid_original"],
+            children=children,
+        )
```

Comment:
I think this **is** the "primary object" -- above we've constructed its children, but this is the root node of the tree that we were actually asked to construct.  Because this is a recursive function, it can't tell what level in the overall construction its operating at.  This object is the current node being constructed.  The children are each also constructed by this -- but in a deeper call to `from_exploded_meta()`